### ISSUE TITLE
Cypress hangs in Azure when running headless in 7.0+

### ISSUE THREAD

**Issue by mmonteiroc** [cypress-io/cypress#15973]
### Current behavior
So we migrated from cypress 6.8.0 to cypress 7.1.0 fixing all breaking changes, once all fixed, we started executing cypress in pipelines, in which they have a limit time execution of 2hours, after they disconnect.

The thing is, we have a suite of 7 smoke tests which they take arround 6-7 mins, but since the migration to the new vesion we get sometimes a tests of them which gets blocked on the pipeline inifinitely until the pipeline it self disconnets by timeout.

Before this migration, this never happened with this test... this is a simple test which just tests that the page is there... nothing of logic inside.

### Desired behavior
To not take inifinite time on a test which before the migration was not happening

### Test code to reproduce
Cannot give the code, but anyways... is not a only test that gets blocked.... its randomly through the tests

### Versions
Last working 6.8.0
Current version 7.1.0

**Comment by mmonteiroc** [comment]
![image](https://user-images.githubusercontent.com/28816737/114667885-52399b80-9d00-11eb-9edf-eedc786fd306.png)

**Comment by mmonteiroc** [comment]
Same test before migration
![image](https://user-images.githubusercontent.com/28816737/114670618-7e0a5080-9d03-11eb-9d5e-2444a9bc4997.png)

**Comment by mmonteiroc** [comment]
same test before migration 
![image](https://user-images.githubusercontent.com/28816737/114670728-9d08e280-9d03-11eb-85ea-ceacd988bb8f.png)

**Comment by mtc3bmtc3b** [comment]
This seems similar to what we're seeing.  Also after a 6.8 -> 7.1 upgrade (saw the same behavior with 7.0.1).  E2E tests that would pass successfully in 6.8 will stall seemingly forever at random places with no indication of any issues.  This happens much more frequently when running headless as part of our CI pipeline, but we do see it when running manually as well.  In the pipeline, we're running Electron and manually it is Chrome.

**Comment by mmonteiroc** [comment]
For us we only see it on headless and only on the pipelines CI (don't know if might be related on the difference of power between pipelines and dev laptops...) 

Also from 6.8.0 to 7.1.0 I noticed a downgrade on the speed of the tests.. tests that would take 20 seconds taking 2 minutes

**Comment by agg23** [comment]
The performance issue is discussed in #15853.

Just to clarify, you have not seen the hang occur locally in `open` mode? Have you tried running the tests locally in `run` mode? It would be nice to see what the command log is displaying during the hang (if Cypress recorded a video of this, that would also be helpful).

**Comment by mmonteiroc** [comment]
So we dont use videos of cypress, so unfortunatelly we don't have that... 
Also this only happens for me in cypress run mode.... with the non-headless mode does not happen....


And again, for me at least only happened on the pipelines, that's why I mean that maybe could be related to the performance of the device running the tests ???? Don't know honestly

**Comment by agg23** [comment]
Have you selected a browser other than Electron in CI (Electron is the default)? Can you try using `open` mode with `--browser electron` (so run something like `cypress open --browser electron`)? Run tests a few times if possible to see if you can have the issue occur.

**Comment by mmonteiroc** [comment]
So in CI we use directly the default one, from the UI we use the default as well, don't know if its other than electron. Will try and tell you if this issue persists.

**Comment by mmonteiroc** [comment]
So executing this command `npx run cypress open --browser electron` does not work for me at all.... I don't know if I might be missing something ???????? Without the electron browser flag works well...
It just print this in terminal and then it keeps there in "starting" for a while and nothing happens
![image](https://user-images.githubusercontent.com/28816737/114729264-1a524880-9d40-11eb-8d03-c268c9cc8222.png)

**Comment by agg23** [comment]
Interesting... I notice you're using Windows. Are your CI machines also Windows?

Try running `DEBUG=* npx run cypress open --browser electron` and post the logs here.

**Comment by mmonteiroc** [comment]
Yes, CI using windows as well.
![image](https://user-images.githubusercontent.com/28816737/114729823-9fd5f880-9d40-11eb-841c-b1cfb118d32e.png)

**Comment by mmonteiroc** [comment]
Yes, CI Using Windows as well 
![image](https://user-images.githubusercontent.com/28816737/114729961-baa86d00-9d40-11eb-90ef-8b7662787245.png)

**Comment by agg23** [comment]
Ah sorry. Noticed you were using Windows and in the same comment gave you a command that won't work in Windows.

Add a entry to your `package.json` `scripts` section:

```json
"scripts": {
  "cypress": "cross-env DEBUG=* cypress open --browser electron"
}
```

Make sure you have `cross-env` (and `cypress`) installed, and run it with `npm run cypress` (or `yarn run cypress`, if you're using `yarn`).

**Comment by mmonteiroc** [comment]
I installed cross-env globally, hope this is as expected.
So like this works properly running with the flag of electron.... the thing is that for confidential reasons, I cannot share with you all the logs that this command generated unfortunatelly as there is somethings that I can't show :(.

But if you ask me to search for something specific I will be happy to do and share only this specific thing

**Comment by agg23** [comment]
> So like this works properly running with the flag of electron....

So it didn't hang this time? Setting the `DEBUG` envvar shouldn't change the program's operation whatsoever. If it truly didn't hang, you should try changing your new script in `package.json` to be just `cypress open --browser electron` and see if you can repro without using `npx`.

**Comment by mmonteiroc** [comment]
but the thing is that for me ONLY hangs when running the tests the following way: `npm run cy:run` and the script contains the follwoing : `"cy:run": "cypress run"`

Which you tell me that the cypress run uses electron by default.... By running cypress open (before using the flag electron) it never hang like this......

Only in headless mode

**Comment by mtc3bmtc3b** [comment]
I'm on Linux with our CI (using electron and cypress run).  One thing I noticed is that with video enabled, the CPU utilization was quite high, though without video enabled it was reasonable.  I also noticed that, regardless of whether video was enabled, we had one Cypress process using 46.8GB of virtual memory.  Running on windows via cypress open shows it also seems much higher in memory usage than with 6.8.0, but not that much higher.  I don't know what 6.8.0 looked like in memory and hard to go back and figure it out as we moving from route() to intercept() as part of this.

**Comment by elevatebart** [comment]
We updated electron in version 7 to get rid of some blocking bugs. It might have caused some of the performance regressions you are seeing.

Thank you for all the details.
We are looking into it as we speak.

**Comment by mmonteiroc** [comment]
Any info on this issue ? At our side we have to downgrade to 6.9.1 as this was causing us really un-stable tests :(

**Comment by elevatebart** [comment]
I fixed #15976 in this PR #16022

Essentially it is impossible to run `cypress run-ct` on windows in 7.1.0. The run mode will wait forever for the Application Under Test to be mounted, but will never respond (bad URL)

7.1.1 fixes the issue.

**Comment by mmonteiroc** [comment]
But so this issue is fixed ???? (i don't understand if you mean that in your comment above) @elevatebart 

I assume is not fixed, otherwise this issue would be close

**Comment by elevatebart** [comment]
The issue I found was within cypress itself. It was simply highlighted by component testing.
We have to wait for the next delivery of cypress for it to work.

I want to wait for users feedback to close it definitely.

If you want to install it locally, to try it out, you can install the temporary version we store on AWS:

```bash
yarn add https://cdn.cypress.io/beta/npm/7.2.0/appveyor-develop-6c12a6c7b706ee4f708a9d19c62f18cd0838a433/cypress.tgz
```

**Comment by tommasomarchionni** [comment]
I resolved the same problem in Azure pipeline, reducing screen size to 1024x720. See [Set screen size when running headless](https://docs.cypress.io/api/plugins/browser-launch-api#Set-screen-size-when-running-headless) for details.
From Cypress 7.0.0 the default screen size is 1980x1080, with this resolution Azure pipeline hangs for hours...

**Comment by mmonteiroc** [comment]
Interesting... we had it fixed in the configuration at this size....
![image](https://user-images.githubusercontent.com/28816737/115565031-5003ae00-a2b9-11eb-8acf-181242670ef0.png)

And was working before 7.1.0... Havent tried with the one you recommend as we don't want to test our app with such small screens ....

We want to be able to test it minimum with the one of the image

**Comment by tommasomarchionni** [comment]
In our Azure pipeline the problem was video recording at 1980x1080: the tests produces video with blocked frames and tests randomly fails or run inifinitely until the pipeline it self disconnets by timeout. You can have viewport larger than your window screen, they are different concepts, this is a screen with 1300x858 viewport and 1024x720 screen
![image](https://user-images.githubusercontent.com/8122764/115568111-44fe4d00-a2bc-11eb-908d-4b4208e631df.png)

**Comment by mmonteiroc** [comment]
Ohh I understand !!!!, so you say that doing this with azure (we have the same) you managed to keep at version 7.0.0+ (or bigger)
Without having any block infinite until the pipeline disconnects ?
Also, appart of using azure, do you run them in a step with linux ? or use powershell and windows ?
@tommasomarchionni

**Comment by mmonteiroc** [comment]
Will try that, but I don't have to much hope as we don't record videos in our pipelines, only screenshots... but Thanks for the tip.... @elevatebart  do you think this might be the cause of the issue ?? the size of the screen ?

**Comment by tommasomarchionni** [comment]
Now I'm using Cypress 7.1.0 in Azure pipeline without problem with this screen size. I use yaml pipeline with pool `windows-2019` and i run cypress test with `script` task.

**Comment by mmonteiroc** [comment]
thanks ! will try and tomorrow after a few pipelines executions will post my feedback here to see if the issue persists, or if this workaround fixes my issue :)

**Comment by elevatebart** [comment]
@mmonteiroc,

I am afraid you will have to wait for 7.1.1 to use Component Testing on windows.
We release it next Monday.

Regarding e2e tests, it should be working in 7.1.0. Have you thought of clearing the cached dependencies?

**Comment by mmonteiroc** [comment]
@elevatebart 

Who said anything about Component Testing ?? I said that we don't do this kind of tests yet.... Our issue is of the infinite block on the smoke && normal E2E tests in the pipelines... 
The most simple tests that in version 6.8.0 were taking 5 secs, they get randomly blocked in the 7.1.0 until my pipeline agent decides to disconnect itself because the timeout of 2 hours we set

And it does happen in the version 7.1.0, the only way of fixing it is rolling back to 6.8.0 or will try the tip of @tommasomarchionni

**Comment by elevatebart** [comment]
Sorry @mmonteiroc 

It was my mistake, the thread got a little long. 
I got your issue mixed up with another fix I made (that has the same symptoms). 
You did not say anything about CT, nobody said anything about CT. 

Regarding this fix, I believe we are still investigating.
I am going to ask.

**Comment by mmonteiroc** [comment]
No problems, I though though that I was getting crazy.

I see you unasigned yourself to this issue, is someone else continue to investigate this ???

At our side we already merged the changes of the screen to see if this unblock us with 7.1.0, tomorrow morning I should have some feedback

**Comment by Bkucera** [comment]
in 7.0.0 we changed the default headless browser resolution to 1920x1080 from 1280x720. based on what @tommasomarchionni has mentioned this could be causing the slow down / crashing. 
Please try reverting back to 1280x720 and seeing if run times go back to normal, here's an example of how to do that
```ts
/**
 * @type {Cypress.PluginConfig}
 */
module.exports = (on, config) => {
  on('before:browser:launch', (browser, options) => {
    return useFixedBrowserLaunchSize(browser, options, config)
  })
})

function  useFixedBrowserLaunchSize (browser, options) {
    if (browser.family === 'firefox') {
      // (height must account for firefox url bar, which we can only shrink to 1px)
      options.args.push(
        '-width', '1280', '-height', '721',
      )
    } else if (browser.name === 'electron') {
      options.preferences.width = 1280
      options.preferences.height = 720
    } else if (browser.family === 'chromium') {
      options.args.push('--window-size=1280,720')
    }
  return options
}

```

**Comment by mmonteiroc** [comment]
Hello ! FYI I did this fix and the tests still hang out....... So this solution dind't solve my issue !

Also, the tests which passed we saw a downgrade in performance even with this fix... this is already mentioned in this issue #15853
But I wanted to mention this, so you know that this for me at least does not fix any of both issues :/

**Comment by mmonteiroc** [comment]
Any news on this issue ? :)

**Comment by jennifer-shehane** [comment]
We made some improvements in performance with our 7.2.0 release. We recommend trying to upgrade to that version. This may not fix all performance issues for everyone however and we're still doing some investigation. 

Please let us know if updating to 7.2.0 improves your run.

**Comment by mmonteiroc** [comment]
Thanks for your feedback @jennifer-shehane, we haven't tried yet 7.2.0 in our CI agents, will try it now and give you my feedback asap I have the results of several runs on our CI
(will try it with the code proposed above, so that the screen size is set before testing - this didn't work in 7.1.0 for us)

However I tried 7.2.0 locally, and even there is an improvement in performance from 7.1.0, there is still some performance regression compared to 6.8.0....(tests that takes 10 min, in 7.2.0 are taking 12/14 minute depending - locally)
 I know this is not related to this thread, but since you commented 7.2.0 improvements in performance, I wanted to let you know

**Comment by mmonteiroc** [comment]
@jennifer-shehane 

We have merged version 7.2.0 + the code to fix the size of the screen provided above in this thread.

So far so good, we are not experiencing for now any blocking infinite (we have executed normal tests more than 5 times, all the same results and good performance),

 BUT our issue initially was in the smoke tests and randomly (for now with this new version we executed it once only) so maybe would be better if we keep this issue open at least until tomorrow. 

So at our side we can test it in the CI agents several times, and verify that indeed version 7.2.0 fixes our issue

**Comment by jennifer-shehane** [comment]
@mmonteiroc Yes, we'd love to hear more feedback on whether 7.2.0 and/or reducing the screen size improves performance in Windows/Azure.

**Comment by mmonteiroc** [comment]
Hello @jennifer-shehane thanks :=) 

We have been executing several smoke tests after a lot of merges (and normal ones as well), seems like the blocking is not happening anymore. COOL !

The only think is that our side we don't know if it's because of the 7.2.0 alone that fixed this, or the mix between 7.2.0 + Screen size....

We know for sure that screen size alone is not, as in 7.1.0 was not working this workaround for us :)

**Comment by mmonteiroc** [comment]
Do you need / want something to us do, so helping finding the actual cause of it ? (can't give reproduce code unfortunately)

**Comment by jennifer-shehane** [comment]
We're glad to hear the hanging is resolved. We just want to make sure performance issues are resolved from the 7.2.0 release. 

The increase in screen size from 7.0 will result in slower runs and we're going to have to come up with a better solution for people so that they can easily get high resolution videos without prohibitive performance.
